### PR TITLE
[Prototype] Use UIA notifications for text output

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
This change makes Windows Terminal raise a `RaiseNotificationEvent()` ([docs](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.automation.peers.automationpeer.raisenotificationevent?view=winrt-22000)) for new text output to the buffer.

This is intended to help Narrator identify what new output appears and reduce the workload of diffing the buffer when a `TextChanged` event occurs.

## Detailed Description of the Pull Request / Additional comments
The flow of the event occurs as follows:
- `TextBuffer::WriteLine()`
   - New text is output to the text buffer. Notify the renderer that we have new text (and what that text is).
- `Renderer::TriggerNewTextNotification()`
   - Cycle through all the rendering engines and tell them to notify handle the new text output.
   - None of the rendering engines _except_ `UiaEngine` has it implemented, so really we're just notifying UIA.
- `UiaEngine::NotifyNewText()`
   - Concatenate any new output into a string.
   - When we're done painting, tell the notification system to actually notify of new events occurring and clear any stored output text. That way, we're ready for the next renderer frame.
- `InteractivityAutomationPeer::NotifyNewOutput()` --> `TermControlAutomationPeer::NotifyNewOutput`
   - NOTE: these are split because of the in-proc and out-of-proc separation of the buffer.
   - Actually `RaiseNotificationEvent()` for the new text output.

## Test cases
- [ ] Base case: "echo hello"
- [ ] Partial line change
- [ ] Passwords (needs Automation Client side work)
- [ ] Scrolling (should be unaffected)
- [ ] Large output

### CODE REVIEW

**Comment by codeofdusk** [comment]
This definitely opens some exciting possibilities, but will need to be handled with care on the client side!

How can/should UIA clients know that notifications are supported and there's no need to use the old strategy?

**Comment by carlos-zamora** [comment]
> This definitely opens some exciting possibilities, but will need to be handled with care on the client side!
> 
> How can/should UIA clients know that notifications are supported and there's no need to use the old strategy?

Could you do something like "if we received a notification event with the ID 'TerminalTextOutput', this is the newest iteration"?

**Comment by codeofdusk** [comment]
> Could you do something like "if we received a notification event with the ID 'TerminalTextOutput', this is the newest iteration"?

Maybe? @josephsl, @leonardder, @michaeldcurran, @jcsteh is it possible to change the overlay class associated with an NVDA object in response to receiving a particular event? (I'm thinking not, but being proven wrong would ease this).

**Comment by jcsteh** [comment]
No. An overlay has to be determined at the time the object is created. That
could be due to a notification event, but it could just as well be due to
some other event; e.g. focus.

That said, you could set a flag if you see such an event. That's not as
clean, but life isn't clean. :)

**Comment by leonardder** [comment]
I see that the terminal edit control currently doesn't have an automation ID. I guess one could be added at this point to distinguish the old from the new implementation.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 1]
```diff
@@ -142,6 +143,24 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+        }
+        dispatcher.RunAsync(Windows::UI::Core::CoreDispatcherPriority::Normal, [weakThis{ get_weak() }, notificationText{ hstring(newOutput) }]() {
+            if (auto strongThis{ weakThis.get() })
+            {
+                strongThis->RaiseNotificationEvent(AutomationNotificationKind::ActionCompleted,
+                                                   AutomationNotificationProcessing::ImportantAll,
+                                                   notificationText,
+                                                   L"TerminalTextOutput");
+            }
+        });
```
Comment (lhecker) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:161]: Don't you need to react to back-pressure here? What if you're called really fast on a slow system? Won't the background jobs just pile up more and more?

File: src/buffer/out/textBuffer.cpp [hunk 2]
```diff
@@ -388,6 +388,7 @@ OutputCellIterator TextBuffer::WriteLine(const OutputCellIterator givenIt,
     const auto written = newIt.GetCellDistance(givenIt);
     const Viewport paint = Viewport::FromDimensions(target, { gsl::narrow<SHORT>(written), 1 });
     _NotifyPaint(paint);
+    _renderTarget.TriggerNewTextNotification(givenIt->Chars());
```
Comment (lhecker) [inline src/buffer/out/textBuffer.cpp:391]: FYI: Considering that this code path (not this method in particular) is the single largest inhibitor for performance in Windows Terminal (slow down of up to 900x compared to other popular terminals), I'm scared of adding things here, because they'll have to leave at some point in the future.

So where should you raise this notifications? Technically - in my personal opinion - you should do so in your Present() method of a IRenderingEngine, because that one is only called once per v-sync, whereas this function can (well... "should") be called thousands of times per second.

Basically diffing your viewport is the fastest approach for us to figure out if something changed and it's basically cost-free too, _unless_ you need to actually tell narrator every line that's being written, but I doubt that's what someone would want... ("Oh no! I accidentally ran `cat lord_of_the_rings.txt` and it won't finish within the next 14 hours!")

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 1 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:161]: We need the jobs to pile up. The screen reader wants to receive the entirety of the output.

Converting this thread to track synchronous notification dispatching. There's a bit of a concern here that the notification events being dispatched in an async way may be processed out of order.

File: src/buffer/out/textBuffer.cpp (hunk 2 above)
Comment (carlos-zamora) [reply src/buffer/out/textBuffer.cpp:391]: I'll have to do a performance comparison for screen reader and non-screen reader scenarios. That's a really good callout.

Discussed offline that I _could_ move all of the `EndPaint` stuff over to `Present`. It wouldn't fix anything here, but if I'm not using the lock in `EndPaint`, it makes sense to. I'll probably do that for this event and move the other ones in a follow-up.

> unless you need to actually tell narrator every line that's being written.

That's the problem right there. All of the new text needs to be presented to the attached screen reader.

As a separate issue though, we need to figure out a way to reduce the frequency of these events, because we're overloading the attached screen reader sometimes. That's a separate PR that I'll be working on.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 3]
```diff
@@ -142,6 +143,24 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+        auto dispatcher{ Dispatcher() };
+        if (!dispatcher)
+        {
+            return;
+        }
+        dispatcher.RunAsync(Windows::UI::Core::CoreDispatcherPriority::Normal, [weakThis{ get_weak() }, notificationText{ hstring(newOutput) }]() {
+            if (auto strongThis{ weakThis.get() })
+            {
+                strongThis->RaiseNotificationEvent(AutomationNotificationKind::ActionCompleted,
+                                                   AutomationNotificationProcessing::ImportantAll,
```
Comment (codeofdusk) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:157]:
```suggestion
                                                   AutomationNotificationProcessing::All,

```

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 1 above)
Comment (lhecker) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:161]: > We need the jobs to pile up.

This could be a matter of semantics, but to be specific, I think "piling them up" is not a good idea and needs to be prevented in this PR.

Windows Terminal can write the famous `big.txt` within half a second, but it has 128457 lines of text. Do we really want 128457 asynchronous jobs to be put into the job queue? This is why I mentioned earlier we need to control for back pressure. Personally I can imagine two solutions:
* A semaphore to let it run ahead by let's say one viewport height
* Only write the next line to the screen when the previous one finished being narrated (a semaphore of size 1)

I can imagine that a user of a narrator might want to have it finish the execution as fast as possible, but I'm worried that creating 128457 isn't a viable solution either.

File: src/buffer/out/textBuffer.cpp [hunk 4]
```diff
@@ -388,7 +388,6 @@ OutputCellIterator TextBuffer::WriteLine(const OutputCellIterator givenIt,
     const auto written = newIt.GetCellDistance(givenIt);
     const Viewport paint = Viewport::FromDimensions(target, { gsl::narrow<SHORT>(written), 1 });
     _NotifyPaint(paint);
-
```
Comment (carlos-zamora) [inline src/buffer/out/textBuffer.cpp:391]: I can undo changes to this file

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 5]
```diff
@@ -142,6 +143,28 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
 
+    void TermControlAutomationPeer::NotifyNewOutput(const std::wstring_view newOutput)
+    {
+        auto dispatcher{ Dispatcher() };
+        if (!dispatcher)
+        {
+            return;
+        }
+
+        dispatcher
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:154]: I should add a comment here pointing out that we're using `.get()` to make this sequential.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 6]
```diff
@@ -142,6 +143,28 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+            .RunAsync(Windows::UI::Core::CoreDispatcherPriority::Normal, [&]() {
+                try
+                {
+                    RaiseNotificationEvent(AutomationNotificationKind::ActionCompleted,
+                                           AutomationNotificationProcessing::All,
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:159]: Should probably also add a note here saying to use `All` to allow for keyboard event interruptions.

File: src/renderer/uia/UiaRenderer.cpp [hunk 7]
```diff
@@ -223,6 +234,26 @@ CATCH_RETURN();
...
+
+// Routine Description:
+// - Used to perform longer running presentation steps outside the lock so the
+//      other threads can continue.
+// - Not currently used by UiaEngine.
+// Arguments:
+// - <none>
+// Return Value:
+// - S_FALSE since we do nothing.
+[[nodiscard]] HRESULT UiaEngine::Present() noexcept
```
Comment (carlos-zamora) [inline src/renderer/uia/UiaRenderer.cpp:256]: The diff is weird here. I basically just moved all of the contents of `EndPaint()` over to `Present()`. This way we're not doing this under lock. Leonard said that's the main difference between the two functions.

Comment (lhecker) [reply src/renderer/uia/UiaRenderer.cpp:256]: It's also necessary in this case, because `RaiseNotificationEvent` calls back into us (`UiaTextRangeBase::GetAttributeValue` specifically) which locks the console. This only works if we don't hold the lock while calling `RaiseNotificationEvent`.

**Comment by carlos-zamora** [comment]
> I see that the terminal edit control currently doesn't have an automation ID. I guess one could be added at this point to distinguish the old from the new implementation.

@codeofdusk Our concern with the automation ID approach is that a user could have multiple terminals open within the same instance (via panes or tabs), so each pane's `TermControl` would need a different ID, which kind of defeats the point.

How does this sound:
- Windows Terminal: update `TermControl`'s `ClassName` from `TermControl` to `TermControl 2`
- Conhost: possibly set/update that class name to `Console 2` (once notification events are implemented there)

That way you also get two different versioning systems in one go.

**Comment by carlos-zamora** [comment]
Large output is acting a little weird in this latest build. Printing 5 lines of text works fine with...
```
for($i=0; $i -lt 5; $i++){echo $i}
```

But printing 10 lines of text with...
```
for($i=0; $i -lt 10; $i++){echo $i}
```
is inconsistent. Occasionally a random number is missing. More often, the new prompt also isn't read.

I'm still debugging through this issue, so if anybody has any ideas, let me know. Might have to do with `dispatcher` in `NotifyNewOutput` being null?

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 1 above)
Comment (miniksa) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp]: Would it be acceptable to make it read out something like "128,000 lines skipped"?

Comment (codeofdusk) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp]: How would an out-of-band announcement like that distinguish itself from output text?

Comment (miniksa) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp]: Good question... I'm just throwing ideas at the wall and seeing if anything sticks. Reading 128,000 lines seems intense.

**Comment by carlos-zamora** [comment]
> Large output is acting a little weird...

This might be fixed now? At least on my end. The cause was that while `UiaRenderer` was sending `_newOutput` to UIA, `_newOutput` would continue to be modified, then cleared at the end of `Present()`. That portion that was added between sending it to UIA and clearing `_newOutput` was never sent to UIA.

The fix was to introduce `_queuedOutput` as a way to copy `_newOutput` over before sending it to UIA, and clear `_newOutput`, thus making it ready for the next `NotifyNewText()` call. The fix was verified to work using accevent.

HOWEVER, Narrator is still cutting off text.

**Comment by josephsl** [comment]
Hi, which notification processing instruction is the event using? I can see that Narrator (and other screen readers) can cut off speech if it receives UIA notification with an “important” processing instruction. Thanks.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 8]
```diff
@@ -142,9 +143,34 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+    }
+
     hstring TermControlAutomationPeer::GetClassNameCore() const
     {
-        return L"TermControl";
+        // After major UIA changes, we need to update the name.
+        // This is a simple way to let screen readers know what "version"
+        // of UIA they are interacting with (i.e. notifications supported).
+        return L"TermControl2";
     }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:174]: @codeofdusk I updated the class name for you here 😊

**Comment by carlos-zamora** [comment]
Ok, seems like Narrator is now reading off large output just fine. I also managed to get rid of the local echo. So I'm taking this PR out of draft.

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 9]
```diff
@@ -1002,6 +1002,12 @@ void Terminal::_WriteBuffer(const std::wstring_view& stringView)
         _AdjustCursorPosition(proposedCursorPosition);
     }
 
+    // Notify UIA of new text.
+    // It's important to do this here instead of in TextBuffer, because here you have access to the entire line of text,
+    // whereas TextBuffer writes it one character at a time via the OutputCellIterator.
+    static std::wstring newline{ L"\n" };
+    _buffer->GetRenderTarget().TriggerNewTextNotification(std::wstring{ stringView } + newline);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalCore/Terminal.cpp:1009]: ### Fun fact

Found some interesting behavior here when playing with different shells...
| Shell | Output when you type <kbd>a</kbd> + <kbd>b</kbd> + <kbd>c</kbd> |
|--|--|
| Ubuntu | `a` ... `b` ... `c` |
| PowerShell | `a` ... `ab` ... `abc` |
| PowerShell (disabled PSReadLine) | `a` ... `b` ... `c` |

As shown above, PowerShell actually rewrites the entire input buffer with each keyboard event! Why? Idk. But that made it particularly frustrating when I was trying to suppress the "local echo".

Fortunately, it seems like PowerShell has been aware of this issue for some time, because _if PowerShell detects that you are using a screen reader, it disables PSReadLine_!

So in your local testing, make sure you open your screen reader (or accessibility insights) first, _then_ open a PowerShell instance. That way, PowerShell will detect that you are using a screen reader and disable PSReadLine automatically (as well as output some text saying that it's done just that).

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 10]
```diff
@@ -28,6 +28,32 @@ namespace XamlAutomation
     using winrt::Windows::UI::Xaml::Automation::Provider::ITextRangeProvider;
 }
 
+static constexpr wchar_t UNICODE_NEWLINE{ L'\n' };
+
+static inline std::wstring Sanitize(std::wstring_view text)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:33]: add some doc comments foo'. What is this sanitizing? (It's stripping all control chars except `\n`, but let the reader know that!

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 11]
```diff
@@ -68,6 +95,14 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         _contentAutomationPeer.SetControlPadding(padding);
     }
 
+    void TermControlAutomationPeer::RecordKeyEvent(const WORD vkey)
+    {
+        if (const auto charCode{ MapVirtualKey(vkey, MAPVK_VK_TO_CHAR) })
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:100]: something tells me there will be future bugs here related to unicode input, but this is probably fine for now. 

My memory harkens back to `Terminal::_CharacterFromKeyEvent`

File: src/cascadia/TerminalControl/TermControlAutomationPeer.h [hunk 12]
```diff
@@ -64,6 +65,7 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         void SignalSelectionChanged() override;
         void SignalTextChanged() override;
         void SignalCursorChanged() override;
+        void NotifyNewOutput(std::wstring newOutput) override;
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.h:68]: make this take a string view?

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 11 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:100]: oh but there's the `towupper` thing below... That might be more complicated than the `Terminal` version. Still, probably good enough for now.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 13]
```diff
@@ -28,6 +28,32 @@ namespace XamlAutomation
...
+    std::for_each(text.begin(), text.end(), [&sanitized](auto&& c) {
+        if (c >= UNICODE_SPACE || c == UNICODE_NEWLINE)
+        {
+            sanitized += c;
+        }
+    });
+    return sanitized;
+}
+
+static constexpr bool IsReadable(std::wstring_view text)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:45]: here too, doc comment would be useful

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Okay 99% of this is super nitpicky but I'm blocking on the string copy thing. That reads like it would still hurt the non-UIA case, which would be sad. Otherwise, I wouldn't block any of those nits.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 10 above)
Comment (lhecker) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:33]: This function should be either static or inline, but not both. static means "local to this .cpp file" and "inline" means "deduplicate functions with the same name". I'd use static.

Additionally you should avoid pessimizing your code.
Your code currently assumes that the input string has a high amount of control characters we need to optimize for. After all you don't reserve() memory for the sanitized string and push characters one by one into it, which indicates that it's optimized for the case of filtering out a maximum amount of control characters.

If you optimize your code for filtering out control characters sometimes you only need to flip the logic from "create a filtered duplicate" to "create a duplicate which is filtered", because most of the time you don't need to filter anything, turning this into a pure duplication. Here's the code:

```cpp
static std::wstring Sanitize(std::wstring sanitized) // <-- fast memory duplication
{
    // vvv filtering which usually won't touch the string at all vvv
    // (The "erase-remove idiom".)
    sanitized.erase(std::remove_if(sanitized.begin(), sanitized.end(), [](wchar_t c) {
        return c < L' ' && c != L'\n';
    }), sanitized.end());
    return sanitized;
}
```

Lastly the only caller of `Sanitize()` has a fresh, mutable copy of the string.
As such the signature should be `static std::wstring Sanitize(std::wstring& sanitized)`.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 14]
```diff
@@ -142,8 +177,62 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
     }
 
+    void TermControlAutomationPeer::NotifyNewOutput(std::wstring newOutput)
+    {
+        // Try to suppress any events (or event data)
+        // that is just the keypress the user made
+        auto sanitized{ Sanitize(newOutput) };
+        while (!_keyEvents.empty() && IsReadable(sanitized))
+        {
+            if (towupper(sanitized.front()) == _keyEvents.back())
```
Comment (lhecker) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:187]: Do you want to uppercase unicode here? In that case `towupper` won't be accurate, because some languages need more or less code points when transforming to upper case (e.g. ß to SS).
If you only want to convert ASCII you can use `til::toupper_ascii`.

To be honest I don't quite get what `_keyEvents` does here. If I'm the only one, don't mind me, but if others feel the same way, a more comprehensive comment would probably be helpful.

File: src/renderer/uia/UiaRenderer.cpp [hunk 15]
```diff
@@ -223,6 +234,31 @@ CATCH_RETURN();
 // Return Value:
 // - S_OK, else an appropriate HRESULT for failing to allocate or write.
 [[nodiscard]] HRESULT UiaEngine::EndPaint() noexcept
+{
+    // Snap this now while we're still under lock
+    // so present can work on the copy while another
+    // thread might start filling the next "frame"
+    // worth of text data.
+    _queuedOutput = std::move(_newOutput);
```
Comment (lhecker) [inline src/renderer/uia/UiaRenderer.cpp:242]: You can make your engine reuse previously allocated memory, by writing:
```cpp
std::swap(_queuedOutput, _newOutput);
_newOutput.clear();
```

This works because `BeginPaint`, `EndPaint`, etc. aren't called again until the previous `Present()` has finished.

Once you do that the `_newOutput += newText;` will be very performant since memory copies are extremely fast (in the order of >10GB/s).

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 16]
```diff
@@ -28,6 +28,32 @@ namespace XamlAutomation
...
+static constexpr bool IsReadable(std::wstring_view text)
+{
+    for (const auto c : text)
+    {
+        if (c > UNICODE_SPACE)
+        {
+            return true;
+        }
+    }
+    return false;
```
Comment (lhecker) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:54]:
```cpp
return text.find(UNICODE_SPACE) != text.end();
```

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 10 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:33]: > Lastly the only caller of `Sanitize()` has a fresh, mutable copy of the string.
As such the signature could be `static std::wstring Sanitize(std::wstring& sanitized)`.

Should I take it a step further and just do `static void Sanitize(std::wstring& sanitized)`?

Comment (lhecker) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:33]: Either that, or you change `NotifyNewOutput` to accept a `std::wstring_view` instead.

Comment (lhecker) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:33]: > @lhecker Should I take it a step further and just do static void Sanitize(std::wstring& sanitized)?

Oh wait... I just realized I forgot to make the return type a `void`. Good catch. 😅
I intended to write `static void Sanitize(std::wstring& sanitized)`. 🙂

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 16 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:54]: > `return text.find(UNICODE_SPACE) != std::wstring_view::npos;`

wait, isn't that different though? I'm trying to find a character `c > UNICODE_SPACE`. So would this work:
```
    return std::find_if(text.begin(), text.end(), [](wchar_t c) {
               return c > UNICODE_SPACE;
           }) != std::wstring_view::npos;
```

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 14 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:187]: Added the following comment in:
```
                // the key event's character (i.e. the "A" key) matches
                // the output character (i.e. "a" or "A" text).
                // We can assume that the output character resulted from
                // the pressed key, so we can ignore it.
```

**Comment by carlos-zamora** [comment]
I've made all the changes on my end. Just need to test the branch out before committing them. That's why the threads above are marked as resolved but the branch isn't up-to-date.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 16 above)
Comment (lhecker) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:54]: Oh woops. In that case the loop is probably easier to understand than the complex `std::find_if`.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 14 above)
Comment (lhecker) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:187]: Isn't that basically a race condition since the communication with the shell is asynchronous?

Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp:187]: But the key down event gets stored until new output comes out. So doesn't that force it to be synchronized?

File: src/renderer/uia/UiaRenderer.cpp [hunk 17]
```diff
@@ -252,35 +288,34 @@ CATCH_RETURN();
...
+            while (_queuedOutput.size() > sapiLimit)
+            {
+                const auto croppedText{ _queuedOutput.substr(0, sapiLimit) };
+                _dispatcher->NotifyNewOutput(croppedText);
+                _queuedOutput = _queuedOutput.substr(sapiLimit);
+            }
+            _dispatcher->NotifyNewOutput(_queuedOutput);
+        }
+        CATCH_LOG();
+    }
```
Comment (lhecker) [inline src/renderer/uia/UiaRenderer.cpp:308]: Use this to avoid copying the entire output string twice in each iteration:
```cpp
try
{
    static constexpr size_t sapiLimit = 1000;

    std::wstring_view output{ _queuedOutput }; // <-- std::wstring::substr returns a string and not a string_view

    for (size_t offset = 0;; offset += sapiLimit) // <-- only a single substr() needed if we use offset instead
    {
        const auto croppedText = output.substr(offset, sapiLimit); // <-- not copying
        if (croppedText.empty())
        {
            break;
        }

        _dispatcher->NotifyNewOutput(croppedText);
    }
}
CATCH_LOG()
```

(You can drop the if condition at the start.)

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp (hunk 14 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp]: > But what if it gets out of sync? For instance what if the shell prints "hello" and I press "e" simultaneously so that it says "helelo". Wouldn't that result in "hlelo" instead of "hello"?

Hmm... I'm not sure if that would happen. But also, keyboard events interrupt the new output from being read, because that's what the screen reader just outright does. So, if that happens, I think it would immediately get interrupted by you pressing the 'e' key and the screen reader would just read 'e'.

(let me know if I'm not clear, it's a bit weird to explain haha)

Comment (codeofdusk) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp]: > keyboard events interrupt the new output from being read

By default, and in 99% of cases, that's true, but this can technically be disabled: disable "speak typed characters", "speak typed words", and "speech interrupt for typed characters" in NVDA's keyboard settings (<kbd>Ctrl+nvda+k</kbd>).

**Review by zadjii-msft — APPROVED** [review]

File: src/renderer/uia/UiaRenderer.cpp [hunk 18]
```diff
@@ -182,6 +182,20 @@ CATCH_RETURN();
...
 
+[[nodiscard]] HRESULT UiaEngine::NotifyNewText(const std::wstring_view newText) noexcept
+try
+{
+    if (!newText.empty())
+    {
+        static constexpr std::wstring_view newline{ L"\n" };
+        _newOutput.reserve(_newOutput.size() + newText.size() + 1);
+        _newOutput.append(newText);
+        _newOutput.append(newline);
```
Comment (lhecker) [inline src/renderer/uia/UiaRenderer.cpp:193]: IMO this is fine:
```cpp
_newOutput.append(newText);
_newOutput.push_back(L'\n');
```

Normally I'm a sucker for preallocating memory, but in this case the memory allocation will be amortized over time due to our `std::swap` trick. So in general, after a while, this here will practically never reallocate again.

**Review by lhecker — APPROVED** [review]

**Comment by DHowett** [comment]
@msftbot make sure @dhowett signs off

**Comment by DHowett** [comment]
Make sure the commit title and body indicate that this is not a prototype... ;P

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 19]
```diff
@@ -1002,6 +1002,11 @@ void Terminal::_WriteBuffer(const std::wstring_view& stringView)
         _AdjustCursorPosition(proposedCursorPosition);
     }
 
+    // Notify UIA of new text.
+    // It's important to do this here instead of in TextBuffer, because here you have access to the entire line of text,
+    // whereas TextBuffer writes it one character at a time via the OutputCellIterator.
+    _buffer->GetRenderTarget().TriggerNewTextNotification(stringView);
```
Comment (DHowett) [inline src/cascadia/TerminalCore/Terminal.cpp:1008]: It feels weird to reach all the way down through the renderer to tell the TermControlAutomationPeer that there was text.

```mermaid
sequenceDiagram
TerminalControl ->> TerminalCore: Text
TerminalCore ->> Renderer: Text
Renderer ->> DX Engine: Text
Renderer ->> UIA Engine: Text
UIA Engine ->> TCAP: Text
TCAP ->> Narrator: Text
```

File: src/renderer/atlas/AtlasEngine.api.cpp [hunk 20]
```diff
@@ -141,6 +141,11 @@ constexpr HRESULT vec2_narrow(U x, U y, AtlasEngine::vec2<T>& out) noexcept
     return S_OK;
 }
 
+[[nodiscard]] HRESULT AtlasEngine::NotifyNewText(const std::wstring_view newText) noexcept
+{
+    return S_OK;
```
Comment (DHowett) [inline src/renderer/atlas/AtlasEngine.api.cpp:148]: you added a base implementation that returns `S_FALSE`; does atlas need an `S_OK` version?

File: src/host/ScreenBufferRenderTarget.cpp [hunk 21]
```diff
@@ -110,3 +110,13 @@ void ScreenBufferRenderTarget::TriggerTitleChange()
         pRenderer->TriggerTitleChange();
     }
 }
+
+void ScreenBufferRenderTarget::TriggerNewTextNotification(const std::wstring_view newText)
```
Comment (DHowett) [inline src/host/ScreenBufferRenderTarget.cpp:114]: Does conhost need this?

**Review by DHowett — CHANGES_REQUESTED** [review]
If the rest of the team is comfortable with this, let's ship it. It scares me, but it's better than what we have today...

Just confirming: _none of this happens if there's no screen reader enabled, right?_

File: src/renderer/atlas/AtlasEngine.api.cpp (hunk 20 above)
Comment (lhecker) [reply src/renderer/atlas/AtlasEngine.api.cpp:148]: AtlasEngine doesn't inherit from `RenderEngineBase`.

**Review by DHowett — COMMENTED** [review]
Note that @j4james is removing the `IRenderTarget` interface... which may complicate this for you.

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 19 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalCore/Terminal.cpp:1008]: To be fair, the UIA Engine already operated this way. It just didn't know _what_ text was written (only _if_ text was written).

File: src/host/ScreenBufferRenderTarget.cpp (hunk 21 above)
Comment (carlos-zamora) [reply src/host/ScreenBufferRenderTarget.cpp:114]: Not yet! Conhost doesn't even use a UIA engine, so this code never gets called. Bill has expressed interest that Conhost should switch over to this notification system because it would make screen readers' lives so much easier (it removes the need to diff). But it makes sense to test this out on Windows Terminal first, then update Conhost later.

**Review by DHowett — APPROVED** [review]

**Comment by leonardder** [comment]
How are screen reader vendors supposed to find out whether events are supported or not? See my suggestion in https://github.com/microsoft/terminal/pull/12358#issuecomment-1030570292

@jcsteh pointed out in https://github.com/microsoft/terminal/pull/12358#issuecomment-1030406804 that we could do something with a flag as soon as an event is received, but I agree that this is far from clean.

**Comment by DHowett** [comment]
I believe we settled on giving the control in Terminal a name; we can revisit this once we port the changes to conhost :)

**Comment by codeofdusk** [comment]
> I believe we settled on giving the control in Terminal a name

Did we? I thought JAWS had issues with that... CC @carlos-zamora? Did you implement the custom property?
